src/algebra/order/kleene.lean

```diff
@@ -0,0 +1,291 @@
+/-
+Copyright (c) 2022 Siddhartha Prasad, Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Siddhartha Prasad, Yaël Dillies
+-/
+import algebra.order.ring.canonical
+import algebra.ring.pi
+import algebra.ring.prod
+import order.hom.complete_lattice
+
+/-!
+# Kleene Algebras
+
+This file defines idempotent semirings and Kleene algebras, which are used extensively in the theory
+of computation.
+
+An idempotent semiring is a semiring whose addition is idempotent. An idempotent semiring is
+naturally a semilattice by setting `a ≤ b` if `a + b = b`.
+
+A Kleene algebra is an idempotent semiring equipped with an additional unary operator `∗`, the
+Kleene star.
+
+## Main declarations
+
+* `idem_semiring`: Idempotent semiring
+* `idem_comm_semiring`: Idempotent commutative semiring
+* `kleene_algebra`: Kleene algebra
+
+## Notation
+
+`a∗` is notation for `kstar a` in locale `computability`.
+
+## References
+
+* [D. Kozen, *A completeness theorem for Kleene algebras and the algebra of regular events*]
+  (kozen1994)
+* https://planetmath.org/idempotentsemiring
+* https://encyclopediaofmath.org/wiki/Idempotent_semi-ring
+* https://planetmath.org/kleene_algebra
+
+## TODO
+
+Instances for `add_opposite`, `mul_opposite`
```

Comment:
```suggestion
Instances for `add_opposite`, `mul_opposite`, `ulift`
```
Do `prod` and `pi` obviously satisfy these axioms too?

Reply:
In fact, the instances for `prod` and `pi` are already down there.